test/AccessLock.t.sol

```diff
@@ -0,0 +1,494 @@
+
+        uint256 balanceOfAfter0 = MockERC20(Currency.unwrap(currency0)).balanceOf(address(this));
+        uint256 balanceOfAfter1 = MockERC20(Currency.unwrap(currency1)).balanceOf(address(this));
+
+        // Balance of currency0 decreases bc we 1) donate and 2) swap zeroForOne.
+        assertLt(balanceOfAfter0, balanceOfBefore0);
+        // Since the donate amount is small, and we swapped zeroForOne, we expect balance of currency1 to increase.
+        assertGt(balanceOfAfter1, balanceOfBefore1);
+    }
+
+    function test_beforeDonate_modifyPosition_succeedsWithAccessLock(uint128 amount) public {
+        vm.assume(amount != 0 && amount > 10 && amount < Pool.tickSpacingToMaxLiquidityPerTick(60));
+
+        // Add liquidity so there is something to donate to.
+        modifyPositionRouter.modifyPosition(
+            key,
+            IPoolManager.ModifyPositionParams({tickLower: -120, tickUpper: 120, liquidityDelta: 100 * 10e18}),
+            ZERO_BYTES
+        );
+
+        uint256 balanceOfBefore1 = MockERC20(Currency.unwrap(currency1)).balanceOf(address(this));
+        uint256 balanceOfBefore0 = MockERC20(Currency.unwrap(currency0)).balanceOf(address(this));
+
+        donateRouter.donate(
+            key, 1 * 10 ** 18, 1 * 10 ** 18, abi.encode(amount, AccessLockHook.LockAction.ModifyPosition)
+        );
+
+        uint256 balanceOfAfter0 = MockERC20(Currency.unwrap(currency0)).balanceOf(address(this));
+        uint256 balanceOfAfter1 = MockERC20(Currency.unwrap(currency1)).balanceOf(address(this));
+
+        // Should have less balance in both currencies from adding liquidity AND donating.
+        assertLt(balanceOfAfter0, balanceOfBefore0);
+        assertLt(balanceOfAfter1, balanceOfBefore1);
+    }
+
+    function test_beforeDonate_donate_succeedsWithAccessLock(uint128 amount) public {
+        vm.assume(amount != 0 && amount > 10 && amount < uint128(type(int128).max)); // precision
+
+        // Add liquidity so there is a position to receive fees.
+        modifyPositionRouter.modifyPosition(
+            key,
+            IPoolManager.ModifyPositionParams({tickLower: -120, tickUpper: 120, liquidityDelta: 100 * 10e18}),
+            ZERO_BYTES
+        );
+
+        uint256 balanceOfBefore1 = MockERC20(Currency.unwrap(currency1)).balanceOf(address(this));
+        uint256 balanceOfBefore0 = MockERC20(Currency.unwrap(currency0)).balanceOf(address(this));
+
+        // Make the swap amount small (like a NoOp).
+        donateRouter.donate(key, 1 * 10 ** 18, 1 * 10 ** 18, abi.encode(amount, AccessLockHook.LockAction.Donate));
+
+        uint256 balanceOfAfter0 = MockERC20(Currency.unwrap(currency0)).balanceOf(address(this));
+        uint256 balanceOfAfter1 = MockERC20(Currency.unwrap(currency1)).balanceOf(address(this));
+
+        // Should have less balance in both currencies.
+        assertLt(balanceOfAfter0, balanceOfBefore0);
+        assertLt(balanceOfAfter1, balanceOfBefore1);
+    }
```

Comment:
what about some edge casier tests like:
- doing one hook operation, then another - the first hook should no longer have access rights but the new one should
- doing one hook operation, then an operation on a pool with _no hook_ - the first hook should no longer have access rights
- .. im sure we can think of more

Reply:
I'll add a hook accessing a lock and applying deltas when there is no outside locker